src/Libraries/Microsoft.Extensions.AI/Functions/AIFunctionFactory.cs

```diff
@@ -320,40 +318,43 @@ static bool IsAsyncMethod(MethodInfo method)
         /// <summary>
         /// Gets a delegate for handling the marshaling of a parameter.
         /// </summary>
-        private static Func<IReadOnlyDictionary<string, object?>, AIFunctionContext?, object?> GetParameterMarshaller(
+        private static Func<IReadOnlyDictionary<string, object?>, CancellationToken, object?> GetParameterMarshaller(
             JsonSerializerOptions serializerOptions,
-            ParameterInfo parameter,
-            ref bool foundAIFunctionContextParameter)
+            ParameterInfo parameter)
         {
             if (string.IsNullOrWhiteSpace(parameter.Name))
             {
                 Throw.ArgumentException(nameof(parameter), "Parameter is missing a name.");
             }
 
-            // Special-case an AIFunctionContext parameter.
-            if (parameter.ParameterType == typeof(AIFunctionContext))
-            {
-                if (foundAIFunctionContextParameter)
-                {
-                    Throw.ArgumentException(nameof(parameter), $"Only one {nameof(AIFunctionContext)} parameter is permitted.");
-                }
-
-                foundAIFunctionContextParameter = true;
-
-                return static (_, ctx) =>
-                {
-                    Debug.Assert(ctx is not null, "Expected a non-null context object.");
-                    return ctx;
-                };
-            }
-
             // Resolve the contract used to marshal the value from JSON -- can throw if not supported or not found.
             Type parameterType = parameter.ParameterType;
             JsonTypeInfo typeInfo = serializerOptions.GetTypeInfo(parameterType);
+            FromServiceProviderAttribute? fspAttr = parameter.GetCustomAttribute<FromServiceProviderAttribute>(inherit: true);
 
-            // Create a marshaller that simply looks up the parameter by name in the arguments dictionary.
-            return (arguments, _) =>
+            // Create a marshaller for the parameter. This produces a value for the parameter based on an ordered
+            // collection of rules.
+            return (arguments, cancellationToken) =>
             {
+                // If the parameter is [FromServiceProvider], try to satisfy it from the service provider
+                // provided via arguments.
+                if (fspAttr is not null &&
+                    (arguments as AIFunctionArguments)?.ServiceProvider is IServiceProvider services)
```

Comment:
Couldn't we just have `InvokeAsync` accept an optional `IServiceProvider` parameter?

Reply:
This PR is showing how it can be added later, and changing InvokeAsync can't be done later. I also think including IServiceProvider in InvokeAsync is an arbitrary additional piece of information that doesn't belong; nothing about AIFunction is related to DI; that's a higher level concern.